Issue of Boost `std::unary_function` on ARM-based mac with Clang 15

## Settings

I’m installing the master version of dealii (commit c54c9999e563e331fd1f88e2ddfded70949ca8a7) on my mac with M3 chip using candi, found some issue related to `std::unary_function` that wasn’t fixed by #16486.

I’m using Clang 15.0.0 and open-mpi installed by `brew install --cc=clang cmake open-mpi gcc@11` as recommended in the wiki documentation.

```
❯ which mpicxx
/opt/homebrew/bin/mpicxx
❯ mpicxx -v
Apple clang version 15.0.0 (clang-1500.1.0.2.5)
Target: arm64-apple-darwin23.3.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

## Issue

The configuring of dealii succeeded, but when building it, the compiler errors about `std::unary_function`

```
.../candi_install_prefix/tmp/unpack/deal.II-master/bundled/boost-1.70.0/libs/serialization/src/../../../include/boost/container_hash/hash.hpp:130:33: error: no template named 'unary_function' in namespace 'std'; did you mean '__unary_function'?
        struct hash_base : std::unary_function<T, std::size_t> {};
                           ~~~~~^~~~~~~~~~~~~~
                                __unary_function
/Library/Developer/CommandLineTools/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__functional/unary_function.h:46:1: note: '__unary_function' declared here
using __unary_function = __unary_function_keep_layout_base<_Arg, _Result>;
```


## Possible solution

Following https://stackoverflow.com/questions/77133361/no-template-named-unary-function-in-namespace-std-did-you-mean-unary-fun, I added another line
`list(APPEND DEAL_II_DEFINITIONS "_LIBCPP_ENABLE_CXX17_REMOVED_UNARY_BINARY_FUNCTION”)` 
under the 
`list(APPEND DEAL_II_DEFINITIONS "BOOST_NO_CXX98_FUNCTION_BASE")` 
line proposed in #16486, and finally get rid of the error.


BTW, I didn’t use the .dmg pre-compiled version of dealii because I’d like to use more of PETSc’s packages like SuperLU. The installation of .dmg version on Apple Silicon is easy and works fine.

--- a/bundled/boost-1.70.0/include/boost/container_hash/hash.hpp
+++ b/bundled/boost-1.70.0/include/boost/container_hash/hash.hpp
@@ -118,7 +118,7 @@ namespace boost
 {
     namespace hash_detail
     {
-#if defined(_HAS_AUTO_PTR_ETC) && !_HAS_AUTO_PTR_ETC
+#if defined(BOOST_NO_CXX98_FUNCTION_BASE)
         template <typename T>
         struct hash_base
         {

